src/decorators/safe-url.decorator.ts

```diff
@@ -0,0 +1,152 @@
+      name: 'isSafeUrl',
+      target: object.constructor,
+      propertyName: propertyName,
+      options: validationOptions,
+      validator: {
+        validate(value: any, args: ValidationArguments) {
+          if (typeof value !== 'string') return false;
+
+          // Instant rejection of URLs containing the most dangerous characters
+          if (/[<>"'`]/.test(value)) {
+            return false;
+          }
+
+          try {
+            // Decoding URL before checking content
+            const decodedUrl = decodeURIComponent(value);
+
+            // Checking if the decoded URL contains dangerous characters
+            if (/[<>"'`]/.test(decodedUrl)) {
+              return false;
+            }
+
+            const url = new URL(value);
+
+            // Basic protocol check
+            if (!['http:', 'https:'].includes(url.protocol)) {
+              return false;
+            }
+
+            // Block Punycode
+            if (url.hostname.includes('xn--')) {
+              return false;
+            }
+
+            // Blocking credentials in URL
+            if (url.username || url.password || url.href.includes('@')) {
+              return false;
+            }
+
+            // Hostname Validation
+            const hostnamePattern =
+              /^([a-zA-Z0-9-]+\.)*[a-zA-Z0-9-]+\.[a-zA-Z]{2,}$/;
+            if (!hostnamePattern.test(url.hostname)) {
+              return false;
+            }
+
+            // We connect the path and parameters to be checked together
+            const fullPath = `${url.pathname}${url.search}${url.hash}`;
+
+            // Decoding full path before checking
+            const decodedPath = decodeURIComponent(fullPath);
+
+            // List of dangerous patterns
+            const dangerousPatterns = [
+              /<.*>/i, //  HTML Tags
+              /[<>"'`]/, // Dangerous characters
+              /javascript:/i, // JavaScript protocol
+              /data:/i, // Data URI
+              /vbscript:/i, // VBScript
+              /file:/i, // File protocol
+              /alert\s*\(/i, // Alert
+              /on\w+\s*=/i, // Event handlers (onclick, onload, etc.)
+              /%0[ad]/i, // CRLF injection
+              /\.\.\/|\.\.\\|%2e%2e/, // Directory traversal
+            ];
+
+            // Checking if any pattern is present in the decoded path
+            if (
+              dangerousPatterns.some((pattern) => pattern.test(decodedPath))
+            ) {
+              return false;
+            }
+
+            // Additional check for encoded characters
+            const encodedChars = /%(?![0-9a-fA-F]{2})/i;
+            if (encodedChars.test(fullPath)) {
+              return false;
+            }
+
+            return true;
+          } catch (e) {
+            return false;
+          }
+        },
+        defaultMessage(args: ValidationArguments) {
+          return `${args.property} zawiera niedozwolone znaki lub jest nieprawidłowym URL`;
```

Comment:
**nitpick**: Consider providing an English translation for the default error message to cater to a wider audience.